src/rnp/rnpcfg.c

```diff
@@ -0,0 +1,502 @@
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
+ * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include "rnpcfg.h"
+#include <rnpsdk.h>
+#include <../common/constants.h>
+
+#include <sys/types.h>
+#include <sys/param.h>
+#include <sys/stat.h>
+#include <regex.h>
+
+/* SHA1 is now looking as though it should not be used.  Let's
+ * pre-empt this by specifying SHA256 - gpg interoperates just fine
+ * with SHA256 - agc, 20090522
+ */
+#define DEFAULT_HASH_ALG "SHA256"
+
+int
+rnp_cfg_init(rnp_cfg_t *cfg)
+{
+    memset((void *) cfg, '\0', sizeof(rnp_cfg_t));
+
+    return RNP_OK;
+}
+
+int
+rnp_cfg_load_defaults(rnp_cfg_t *cfg)
+{
+    rnp_cfg_setint(cfg, CFG_OVERWRITE, 1);
+    rnp_cfg_set(cfg, CFG_OUTFILE, NULL);
+    rnp_cfg_set(cfg, CFG_HASH, DEFAULT_HASH_ALG);
+    rnp_cfg_set(cfg, CFG_CIPHER, "cast5");
+    rnp_cfg_setint(cfg, CFG_MAXALLOC, 4194304);
+    rnp_cfg_set(cfg, CFG_SUBDIRGPG, SUBDIRECTORY_RNP);
+    rnp_cfg_set(cfg, CFG_SUBDIRSSH, SUBDIRECTORY_SSH);
+    rnp_cfg_setint(cfg, CFG_NUMTRIES, MAX_PASSPHRASE_ATTEMPTS);
+
+    return RNP_OK;
+}
+
+int
+rnp_cfg_apply(rnp_cfg_t *cfg, rnp_params_t *params)
+{
+    int         passfd;
+    const char *stream;
+
+    /* enabling core dumps if user wants this */
+    if (rnp_cfg_getint(cfg, CFG_COREDUMPS)) {
+        params->enable_coredumps = 1;
+    }
+
+    /* checking if password input was specified */
+    if ((passfd = rnp_cfg_getint(cfg, CFG_PASSFD))) {
+        params->passfd = passfd;
+    }
+
+    /* stdout/stderr and results redirection */
+    if ((stream = rnp_cfg_get(cfg, CFG_IO_OUTS))) {
+        params->outs = stream;
+    }
+
+    if ((stream = rnp_cfg_get(cfg, CFG_IO_ERRS))) {
+        params->errs = stream;
+    }
+
+    if ((stream = rnp_cfg_get(cfg, CFG_IO_RESS))) {
+        params->ress = stream;
+    }
+
+    /* detecting keystore pathes and format */
+    if (!rnp_cfg_get_ks_info(cfg, params))
+        return RNP_FAIL;
+
+    /* default key/userid */
+    if (!rnp_cfg_get_defkey(cfg, params))
+        return RNP_FAIL;
+
+    return RNP_OK;
```

Comment:
return true/false or    PGP_E_OK + more descriptive error codes in case of error